facebookresearch/ParlAI
# Conversion of turn annotations to Mephisto

**Patch description**
This PR contains code to run the turn annotations task on Mephisto, including running the parts that get the worker connected to the task, running them through onboarding, loading them into a task world, talking with a model, and saving the results in a format compatible with the old scripts. Some of the code has been adapted from the static version, but it has been updated to use react state where possible instead of relying on any html accessors. Data-saving with this will put the data into the normal Mephisto directories, and a separate file of conversation data will be saved in a backwards-compatible format.

The goal here was to get the flow working, I'm sure that some of the `World` code can be cleaned up, and it's unclear that the `TurkLikeAgent` is even required anymore.

The old `parlai.mturk.tasks.turn_annotations` code has been removed because it is now no longer needed.

**Testing steps**
Ran the run script, ensured that both onboarding and the main world can be run to completion. See `tests/crowdsourcing/tasks/turn_annotations/test_turn_annotations.py` for a test to check the final agent state and the file of final chat data: this required quite a bit of refactoring of the existing `parlai/crowdsourcing` testing code.

The following is a test chat between myself and a classifier model I'm testing on some LIGHT object descriptions.

![Screen Shot 2020-11-06 at 11 51 46 AM](https://user-images.githubusercontent.com/1276867/98845372-9a842300-241b-11eb-87b4-d9204f7c2a98.png)
![Screen Shot 2020-11-06 at 11 54 01 AM](https://user-images.githubusercontent.com/1276867/98845382-9f48d700-241b-11eb-8f7d-5fdeb89bfcce.png)
![Screen Shot 2020-11-10 at 4 17 37 PM](https://user-images.githubusercontent.com/1276867/98845390-a374f480-241b-11eb-9f0e-2416ff3a9345.png)
![Screen Shot 2020-11-10 at 4 17 54 PM](https://user-images.githubusercontent.com/1276867/98845404-a839a880-241b-11eb-883a-c85e8bf05f91.png)
![Screen Shot 2020-11-10 at 4 23 09 PM](https://user-images.githubusercontent.com/1276867/98845420-ac65c600-241b-11eb-8738-54b86e27d965.png)
![Screen Shot 2020-11-10 at 4 23 20 PM](https://user-images.githubusercontent.com/1276867/98845435-b25ba700-241b-11eb-9583-d0d1fd43682a.png)
![Screen Shot 2020-11-10 at 4 33 30 PM](https://user-images.githubusercontent.com/1276867/98845444-b687c480-241b-11eb-8e02-3f990f2f80b4.png)


**Other information**
This PR relies on the following changes to Mephisto in order to function 100% correctly:
- https://github.com/facebookresearch/Mephisto/pull/298
- https://github.com/facebookresearch/Mephisto/pull/310
- https://github.com/facebookresearch/Mephisto/pull/311
